doc(promtail): Doc about how log rotate works with promtail. (#8096)
grafana/loki: docs/sources/clients/promtail/logrotation.md

# Promtail and Log rotation

## Why log rotation matters?

At any point in time, there can be three processes working on a log file as shown in the image below.
![block_diagram](./logrotation-components.png)

1. Appender - A writer that keep appending to a log file. This can be your application or some system daemons like syslog, docker log driver or kubelet etc.
2. Tailer - A reader that read log lines as it is being appended e.g agents like Promtail.
3. Log Rotater - A process that rotates the log either based on time (say scheduled every day) or size (say a log file reached its max size)

NOTE: Here `fd` defines a file descriptor. Once a file is open for read or write, OS gives back unique file descriptor (an ingeter) per process and all the operations like read and write are done over that file descriptor. In other words, once file is opened successfully, file descriptor matters more than the file name.

One of the critical component here is log rotater. Let's understand how it impacts other components like appender and tailer.

> In general, a log can be rotated in two ways.

Author: Good point. Here when we talk about "ways to logrotate", it's about "how" exactly log rotation works rather than "how often" it is rotated. May be need some different wording there. I will rephrase it. thanks :)

![block_diagram](./logrotation2.png)

Both types of log rotation seems to give same end result. However there are some subtle differences.

(1) favours the appender in a sense, that file descriptor it is holding for original log file `error.log` doesn't change. So it can keep writing to same file descriptor, re-opening of the file is not needed.

However (1) has a serious problem if we bring `tailer` into the picture. There is a race between truncating the file and `tailer` finish reading that log file. Meaning, there is a high chance log rotate can truncate the file `error.log` before `tailer` read everything from it.

This is where (2) can help. Here when log file `error.log` is renamed to `error.log.1`, the `tailer` still holding file descriptor of `error.log.1` and it can continue reading the log file until it's completed. But that comes with the tradeoff, with (2) you have to signal the `appender` to reopen `error.log` (and `appender` should be able to reopen it), else it would be keep writing to `error.log.1` (because of same file descriptor). Almost all of the `appender` in real life (e.g: syslog, kubelet, docker log driver) supports reopening log file when it is renamed.

> We recommend (2) as that is the one which works well with Promtail (or any agent) without any dataloss. Now let's understand how do we exactly configure log rotation in different platforms.

Author: Valid concern. AFAIK, well known agents like [Filebeat(elastic)](https://www.elastic.co/guide/en/beats/filebeat/current/how-filebeat-works.html#at-least-once-delivery), vector, datadog agent all works with polling and recommend (1). However I think you are right we don't have to make that assumption considering some may change and new agents may emerge. Will remove it.

## Configure log rotation.

> Your logs can be rotated by different components depending on where you are running your application or services. If you are running on Linux bare metal or VM, high chance you will be using [`logrotate`](https://man7.org/linux/man-pages/man8/logrotate.8.html) utility. However if you are running in kubernetes, it's not that obvious who rotates the logs and interestingly it may dependes on what container runtime your kubernetes cluster is using.

Author: Here Linux bare metal means a single bare metal machine running Linux.

Will change it to `Linux bare metal or Linux VM`.

### Non kubernetes

Usually in Linux bare metal or VM log rotation is handled by [`logrotate`](https://man7.org/linux/man-pages/man8/logrotate.8.html) utility.

> The two different types of log rotation as explained before are supported in `logrotate` as well.

Author: Like "types" better than "alternatives" here :)

1. `containerLogMaxSize` - It is a quantity defining the maximum size of the container log file before it is rotated. For example: "5Mi" or "256Ki". Default: "10Mi"
2. `containerLogMaxFiles` - It specifies the maximum number of container log files that can be present for a container. Default: 5

Both should be part of `kubelet` config. Check cloud provider docs for tweaking this configs in managed version of Kubernetes cluster. Examples [GKE](https://cloud.google.com/kubernetes-engine/docs/how-to/node-system-config#create), [AKS](https://learn.microsoft.com/en-us/azure/aks/custom-node-configuration#use-custom-node-configuration) and [EKS](https://eksctl.io/usage/customizing-the-kubelet/#customizing-kubelet-configuration)

NOTE: Log rotation managed by `kubelet` supports only rename + create (for a good reason) and doesn't support copy + truncate.

If `kubelet` is not configured to manage the log rotation, then it's upto CRI (Container Runtime Interface) the cluster is using or can be managed by `logrotate` utility in the kubernetes node itself.

Check your container runtime on your nodes via
```bash

> $ kubectl get nodes -o wide

Author: Yea thought about that, but didn't do it because, we cannot guarantee the columns order or may add additional columns in future thus breaking our awk magic.

I could have done it something like 
```
k get nodes -o custom-columns=RUNTIME:.status.nodeInfo.containerRuntimeVersion
```
But again, it also feels like complicated. `kubectl get nodes -o wide` seems simple enough to figure out the container runtime.

Out of all these logging drivers only `local` (default) and `json-file` supports log rotation. You configure following `log-opts` under `/etc/docker/daemon.json`

1. `max-size` - The maximum size of the log before it is rolled. A positive integer plus a modifier representing the unit of measure (k, m, or g). Defaults to 20m.
2. `max-file` - The maximum number of log files that can be present. If rolling the logs creates excess files, the oldest file is removed. A positive integer. Defaults to 5.

Example `/etc/docker/daemon.json`
```bash
{
  "log-driver": "local",
  "log-opts": {
    "max-size": "10m",
	"max-file": "10"
  }
}
```

If neither of `kubelet` or `CRI` is configured for log rotate, then we can still use `logrotate` utility on kubernetes nodes for the logration as explained previously.

**Recommended: Use kubelet for log rotation**

## Configure Promtail

Promtail uses `polling` to watch for file changes. And combined with `polling` and [copy and truncate](#copy-and-truncate) log rotation, there can be high chance, some logs may be lost before promtail reads all log lines(as file may got truncated).

> In that case we strongly recommend to change the log rotation strategy to [rename and create](#rename-and-create) for long term solution. For the short term, you can tweak promtail client's `batchsize` [config](https://grafana.com/docs/loki/latest/clients/promtail/configuration/#clients) to higher values (like 5M or 8M) giving more room to read loglines without frequently waiting for push response from Loki server.

Reviewer: ```suggestion
Therefore, for a long-term solution, we strongly recommend changing the log rotation strategy to [rename and create](#rename-and-create). Alternatively, as a workaround in the short term, you can tweak the promtail client's `batchsize` [config](https://grafana.com/docs/loki/latest/clients/promtail/configuration/#clients) to set higher values (like 5M or 8M). This gives Promtail more room to read loglines without frequently waiting for push response from the Loki server.
```

# Promtail and Log Rotation

## Why does log rotation matters?

At any point in time, there may be three processes working on a log file as shown in the image below.
![block_diagram](./logrotation-components.png)

1. Appender - A writer that keeps appending to a log file. This can be your application or some system daemons like Syslog, Docker log driver or Kubelet, etc.

> 2. Tailer - A reader that reads log lines as they are appended e.g agents like Promtail.

Reviewer 2: ```suggestion
2. Tailer - A reader that reads log lines as they are appended, for example, agents like Promtail.
```

> 3. Log Rotater - A process that rotates the log file either based on time (say scheduled every day) or size (say a log file reached its max size).

Reviewer 2: ```suggestion
3. Log Rotator - A process that rotates the log file either based on time (for example, scheduled every day) or size (for example, a log file reached its maximum size).
```

> > **NOTE:** Here `fd` defines a file descriptor. Once a file is open for read or write, OS gives back a unique file descriptor (usually an integer) per process, and all the operations like read and write are done over that file descriptor. In other words, once the file is opened successfully, the file descriptor matters more than the file name.

Reviewer 2: "OS gives back"  This confused me, do you mean "the operating system gives back a unique file descriptor"?  

Also not sure how to give feedback on a graphic in GitHub, but in logrotation-components.png should there also be a `fd` notation on the line between the Log Rotator and the error.log?

Author: > "OS gives back" This confused me, do you mean "the operating system gives back a unique file descriptor"?

Yes.

> Also not sure how to give feedback on a graphic in GitHub, but in logrotation-components.png should there also be a fd notation on the line between the Log Rotator and the error.log?

Technically yes Log Rotator also deals with file descriptor to error.log. But we don't care about how file descriptors are handled in log rotator. That's why left it out.

Reviewer 2: Then I think I would rewrite "the OS gives back" as either 
"The operating system returns a unique file description"   or
"The operating system assigns a unique file description"

Author: Went with "The Operating System returns a unique file description"

> One of the critical components here is the log rotater. Let's understand how it impacts other components like the appender and tailer.

Reviewer 2: Rotat**or** is the noun or adjective form, rotat**er** is the verb form.

> In general, when a log is rotated, it can be done in two ways.

Author: Added this avoid confusing between "how" log is rotated with "how frequent" log is rotated (e.g: daily, weakly, monthly, etc).

Previously when I was using "A log file can be rotated in two ways". The above question was raised. why only two ways? what about montly, daily, etc.

That's why added like this to differentiate that we are talking purely about "how" log is rotated internally. Not "how frequent" log is rotated.

Will leave it like this for now.

> In general, when a log is rotated, it can be done in two ways.

Reviewer 2: This is passive voice, could we make it active to indicate who or what is doing the rotating?  Is this an action that users have to take, or does it happen automatically?

In general, when a software program rates a log, it can be done in two ways.
In general, when Promtail rotates a log, it can be done in two ways.

Author: > This is passive voice, could we make it active to indicate who or what is doing the rotating? Is this an action that users have to take, or does it happen automatically?

It happens automatically (internal to log rotator).

I like "In general, when a software program rotates a log, it can be done in two ways internally."

Reviewer 2: > I like "In general, when a software program rotates a log, it can be done in two ways internally."

Then I'd use that wording.

Author: Nice. updated already :+1:

Given a log file `error.log`

> 1. Copy the log file with a different name e.g `error.log.1` and truncate the original log file `error.log`.

Reviewer 2: ```suggestion
1. Copy the log file with a different name, for example, `error.log.1` and truncate the original log file `error.log`.
```

> 2. Rename the log file with a different name e.g: `error.log.1` and create a new log file with the original name `error.log`.

Reviewer 2: ```suggestion
2. Rename the log file with a different name, for example,  `error.log.1` and create a new log file with the original name `error.log`.
```

> In both cases, after log rotation, all new log lines go to the original `error.log` file.

Reviewer 2: ```suggestion
In both cases, after log rotation, all new log lines are written to the original `error.log` file.
```

> These types are shown in the images below.

Reviewer 2: ```suggestion
These two methods of log rotation are shown in the following images.
```

### Copy and Truncate
![block_diagram](./logrotation-copy-and-truncate.png)

### Rename and Create
![block_diagram](./logrotation-rename-and-create.png)

Both types of log rotation seem to give the same result. However, there are some subtle differences.

(1) favors the `appender` as its descriptor for the original log file `error.log` doesn't change. Therefore, it can keep writing to the same file descriptor. In other words, re-opening the file is not needed.

However (1) has a serious problem if we bring the `tailer` into account. There is a race between truncating the file and the `tailer` finishing reading that log file. Meaning, there is a high chance of the log rotation mechanism truncating the file `error.log` before the `tailer` reads everything from it.

> This is where (2) can help. Here, when the log file `error.log` is renamed to `error.log.1`, the `tailer` still holds the file descriptor of `error.log.1`. Therefore, it can continue reading the log file until it's completed. But that comes with the tradeoff: with (2), you have to signal the `appender` to reopen `error.log` (and `appender` should be able to reopen it). Otherwise, it would keep writing to `error.log.1` as the file descriptor won't change. The good news is that most of the popular `appender` solutions (e.g: Syslog, Kubelet, Docker log driver) support reopening log files when they are renamed.

Reviewer 2: ```suggestion
This is where Rename and Create(2) can help. Here, when the log file `error.log` is renamed to `error.log.1`, the `tailer` still holds the file descriptor of `error.log.1`. Therefore, it can continue reading the log file until it is completed. But that comes with the tradeoff: with (2), you have to signal the `appender` to reopen `error.log` (and `appender` should be able to reopen it). Otherwise, it would keep writing to `error.log.1` as the file descriptor won't change. The good news is that most of the popular `appender` solutions (for example, Syslog, Kubelet, Docker log driver) support reopening log files when they are renamed.
```

> We recommend (2) as that is the one which works well with Promtail (or any similar log scraping agent) without any data loss. Now let's understand how we exactly configure log rotation in different platforms.

Reviewer 2: ```suggestion
We recommend Rename and Create(2) as that is the method which works well with Promtail (or any similar log scraping agent) without any data loss. Now let's understand exactly how we configure log rotation in different platforms.
```

## Configure log rotation.

> Your logs can be rotated by different components depending on where you are running your application or services. If you are running on Linux bare metal or Linux VMs, high chance you will be using [`logrotate`](https://man7.org/linux/man-pages/man8/logrotate.8.html) utility. However, if you are running in Kubernetes, it's not that obvious who rotates the logs and interestingly it may depend on what container runtime your Kubernetes cluster is using.

Reviewer 3: ```suggestion
Your logs can be rotated by different components depending on where you are running your applications or services. If you are running them on Linux bare metal or Linux VMs, there is a high chance you will be using the [`logrotate`](https://man7.org/linux/man-pages/man8/logrotate.8.html) utility. However, if you are running them in Kubernetes, it's not that obvious who rotates the logs and interestingly it may depend on what container runtime your Kubernetes cluster is using.
```

Reviewer 3: Is there a difference between "Linux bare metal" and "Linux VMs"?

### Non-Kubernetes

As mentioned above, in Linux machines, log rotation is often handled by [`logrotate`](https://man7.org/linux/man-pages/man8/logrotate.8.html) utility.

> It supports both types of log rotation explained before.

Reviewer 2: ```suggestion
It supports both methods of log rotation described previously.
```

#### Copy and Truncate
```
/var/log/apache2/*.log {
        weekly
        maxsize 1G
        copytruncate
}
```

Here `copytruncate` mode works exactly like (1) explained above.

#### Rename and Create **(Recommend)**
```
/var/log/apache2/*.log {
        weekly
        maxsize 1G
        create
}
```
Here `create` mode works like (2) explained above. The `create` mode is optional because it's the default mode in `logrotate`.

> The configuration for `logrotate` is usually located in `/etc/logrotate/`.

Reviewer 2: It looks like lines 74 and 75 are describing `logrotate` in general (and apply to both methods), in which case I'd move these up to line 52, above the headings for Copy and Truncate and Rename and Create.  Having them after the Rename and Create heading implies that these two sentences only apply to the Rename and Create method.

> It has a wide range of [options](https://man7.org/linux/man-pages/man8/logrotate.8.html) for compression, mailing, running scripts pre and post-rotation, etc.

Reviewer 3: ```suggestion
It has a wide range of [options](https://man7.org/linux/man-pages/man8/logrotate.8.html) for compression, mailing, running scripts pre- and post-rotation, etc.
```

### Kubernetes

> [Kubernetes Service Discovery in Promtail]({{<relref "./scraping.md">}}#kubernetes-discovery) also uses file-based scraping. Meaning, logs from your pods are stored on the nodes and Promtail scrapes them from those files.

Reviewer 2: ```suggestion
[Kubernetes Service Discovery in Promtail]({{<relref "./scraping.md">}}#kubernetes-discovery) also uses file-based scraping. Meaning, logs from your pods are stored on the nodes and Promtail scrapes the pod logs from the node files.
```

> [Kubernetes Service Discovery in Promtail](https://grafana.com/docs/loki/latest/clients/promtail/scraping/#kubernetes-discovery) also uses file-based scraping. Meaning, logs from your pods are stored on the nodes and Promtail scrapes them from those files.

Reviewer 3: We should use `relref`s for internal links:

```suggestion
[Kubernetes Service Discovery in Promtail]({{<relref "./scraping.md">}}#kubernetes-discovery) also uses file-based scraping. Meaning, logs from your pods are stored on the nodes and Promtail scrapes them from those files.
```

> We can [configure](https://kubernetes.io/docs/concepts/cluster-administration/logging/#log-rotation) the `kubelet` process running on each node to manage log rotation via two configs.

Reviewer 2: ```suggestion
You can [configure](https://kubernetes.io/docs/concepts/cluster-administration/logging/#log-rotation) the `kubelet` process running on each node to manage log rotation via two configuration settings.
```

1. `containerLogMaxSize` - It defines the maximum size of the container log file before it is rotated. For example: "5Mi" or "256Ki". Default: "10Mi".
2. `containerLogMaxFiles` - It specifies the maximum number of container log files that can be present per container. Default: 5

Both should be part of the `kubelet` config. If you run a managed version of Kubernetes in Cloud, refer to your cloud provider documentation for configuring `kubelet`. Examples [GKE](https://cloud.google.com/kubernetes-engine/docs/how-to/node-system-config#create), [AKS](https://learn.microsoft.com/en-us/azure/aks/custom-node-configuration#use-custom-node-configuration) and [EKS](https://eksctl.io/usage/customizing-the-kubelet/#customizing-kubelet-configuration).

> NOTE: Log rotation managed by `kubelet` supports only rename + create (for a good reason) and doesn't support copy + truncate.

Reviewer 2: If you're not going to tell me the reason, I'd just delete "(For a good reason)" in this NOTE.

Author: The reason is what explained above. Why other method `copy  + truncate` is problematic

Reviewer 2: Then just delete "(For a good reason)"

Author: Done.

> NOTE: Log rotation managed by `kubelet` supports only rename + create (for a good reason) and doesn't support copy + truncate.

Reviewer 3: Same as above.

> If `kubelet` is not configured to manage the log rotation, then it's up to the CRI (Container Runtime Interface) the cluster uses. Alternatively, log rotation can be managed by the `logrotate` utility in the Kubernetes node itself.

Reviewer 2: ```suggestion
If `kubelet` is not configured to manage the log rotation, then it's up to the Container Runtime Interface (CRI) the cluster uses. Alternatively, log rotation can be managed by the `logrotate` utility in the Kubernetes node itself.
```

Check your container runtime (CRI) on your nodes by running:
```bash
$ kubectl get nodes -o wide
```

Two of the commonly used CRI implementations are `containerd` and `docker`.

#### containerd CRI

> At the time of writing this guide, `containerd` [doesn't support any way of log rotation](https://github.com/containerd/containerd/issues/4830#issuecomment-744744375). In this case, rotation is often handled by `kubelet` itself. Managed Kubernetes clusters like GKE and AKS use `containerd` as runtime and log rotation is handled by `kubelet`. [EKS after version 1.24](https://docs.aws.amazon.com/eks/latest/userguide/dockershim-deprecation.html) also uses `containerd` as its default container runtime.

Reviewer 2: ```suggestion
At the time of writing this guide, `containerd` [doesn't support any method of log rotation](https://github.com/containerd/containerd/issues/4830#issuecomment-744744375). In this case, rotation is often handled by `kubelet` itself. Managed Kubernetes clusters like GKE and AKS use `containerd` as runtime and log rotation is handled by `kubelet`. [EKS after version 1.24](https://docs.aws.amazon.com/eks/latest/userguide/dockershim-deprecation.html) also uses `containerd` as its default container runtime.
```

#### docker CRI

> When using `docker` as runtime(EKS before 1.24 uses it by default), log rotation is managed by its logging driver(if supported). Docker has [support for several logging drivers](https://docs.docker.com/config/containers/logging/configure/#supported-logging-drivers).

Reviewer 3: ```suggestion
When using `docker` as runtime (EKS before 1.24 uses it by default), log rotation is managed by its logging driver (if supported). Docker has [support for several logging drivers](https://docs.docker.com/config/containers/logging/configure/#supported-logging-drivers).
```

> One can find what logging driver `docker` is using by running the following command

Reviewer 2: ```suggestion
You can determine which logging driver `docker` is using by running the following command:
```

```bash
 docker info --format '{{.LoggingDriver}}'
```

Out of all these logging drivers only the `local` (default) and the `json-file` drivers support log rotation. You can configure the following `log-opts` under `/etc/docker/daemon.json`

1. `max-size` - The maximum size of the log file before it is rotated. A positive integer plus a modifier representing the unit of measure (k, m, or g). Defaults to `20m` (20 megabytes).
2. `max-file` - The maximum number of log files that can be present. If rolling the logs creates excess files, the oldest file is removed. A positive integer. Defaults to 5.

Example `/etc/docker/daemon.json`:

> ```bash

Reviewer 3: ```suggestion
```json
```

{
  "log-driver": "local",
  "log-opts": {
    "max-size": "10m",
    "max-file": "10"
  }
}
```

If neither `kubelet` nor `CRI` is configured for rotating logs, then the `logrotate` utility can be used on the Kubernetes nodes as explained previously.

> **We recommend using kubelet for log rotation.**

Reviewer 2: I would make this a note, bold could be confused with heading formatting.

> **We recommend using kubelet for log rotation**

Reviewer 3: ```suggestion
**We recommend using kubelet for log rotation.**
```

## Configure Promtail

> Promtail uses `polling` to watch for file changes. A `polling` mechanism combined with a [copy and truncate](#copy-and-truncate) log rotation may result in losing some logs. As explained earlier in this guide, this happens when the file is truncated before Promtail reads all the log lines from such a file.

Reviewer 2: ```suggestion
Promtail uses `polling` to watch for file changes. A `polling` mechanism combined with a [copy and truncate](#copy-and-truncate) log rotation may result in losing some logs. As explained earlier in this topic, this happens when the file is truncated before Promtail reads all the log lines from such a file.
```
